**Question 4**

(a) Zipcode of a city (e.g., 48824 for East Lansing).


✅Cannot add or devide, so zipcodes are nominal.

(b) Month of the year, whose value is in the range between 1 (January) to 12 (December).


✅Monthes are categories, thus monthes are nominal.

(c) Speed of a fast-moving object in a medium, measured in terms of its Mach number (see https://en.wikipedia.org/wiki/Mach_number). For example, an object moving at Mach 1 has a speed equal to the speed of sound in the given medium.


✅Speed can be added or devided, has a zero point, so speed is ratio.

(d) Number of daily visitors to a web site.


✅Number of visitors can be added or devided, has zero point, so it is ratio.

(e) Homeland Security Advisory System ratings - code red/orange/etc.


✅Ratings cannot be devided, can be compared and sorted, cannot be subtracted, so ratings are ordinal.

(f) URL of a Web page.


✅URL cannot be added or devided, doesn't have rank, cannot be compared, so URLs are nominal.

(g) TheSaffir-Simpsonhurricanewindscale(seehttps://en.wikipedia. org/wiki/Saffir%E2%80%93Simpson_scale).


✅The scale can be compared but cannot be subtracted, has order, so the scale is ordinal.

(h) The wealth of a nation, measured in terms of gross domestic product (GDP) per capita above the world’s average of $10,880. For example, the wealth for Norway is $91,616 since its GDP per capita is $102,496.

✅GDP can be subtracted, can be compared, thus GPD is ratio.

**Question 5**

**a)**

✅Answer:

SELECT ReleaseYear FROM Movie WHERE Title = Jurassic Park

**b)**

✅Answer:

SELECT FirstName, LastName FROM Person, Movie WHERE Movie(Title) = Jurassic Park

**c)**

✅Answer:

SELECT Title FROM Movie WHERE Director = Steven Spielberg

**d)**

✅Answer:

SELECT COUNT(*) FROM Person, Movie GROUP BY Person(ID)

**Question 6**

**a)**

1. Select the movies that are released in 2007.
2. One column, the movie title.
3. Minimum is 0, Maximum is 150.

**b)**

1. Select the director's name for "Star Wars".
2. Two columns, the person's first name and last name.
3. Minimum is 0, Maximum is 1.

**c)**

1. Select the people who is both an actor and the director in a movie.
2. Two columns, the person's First name and Last Name.
3. Minmimum is 0, Maximum is 200.

**d)**

1. Select unique people who was in multiple movies.
2. Two columns, the person's First name and Last Name.
3. Minimum is 0, Maximum is 200.

**Question 7**

The easiest way to do this should be using an example.

For example, we have a list [1,4,9,13,14,15,16,17,26]

In [42]:
import pandas as pd 
import numpy as np
lis = [1,4,9,13,14,15,16,17,26]
wid = pd.cut(lis,3)
freq = pd.qcut(lis,3)
k = 0
l = 0
for i in wid:
    print("equal width",i,lis[k])
    k+=1
for j in freq:
    print("equal frequency",j,lis[l])
    l+=1


equal width (0.975, 9.333] 1
equal width (0.975, 9.333] 4
equal width (0.975, 9.333] 9
equal width (9.333, 17.667] 13
equal width (9.333, 17.667] 14
equal width (9.333, 17.667] 15
equal width (9.333, 17.667] 16
equal width (9.333, 17.667] 17
equal width (17.667, 26.0] 26
equal frequency (0.999, 11.667] 1
equal frequency (0.999, 11.667] 4
equal frequency (0.999, 11.667] 9
equal frequency (11.667, 15.333] 13
equal frequency (11.667, 15.333] 14
equal frequency (11.667, 15.333] 15
equal frequency (15.333, 26.0] 16
equal frequency (15.333, 26.0] 17
equal frequency (15.333, 26.0] 26


**a)**

In [46]:
lis_a = lis-(np.mean(lis))
wid_a = pd.cut(lis_a,3)
freq_a = pd.qcut(lis_a,3)
k = 0
l = 0
for i in wid_a:
    print("equal width",i,"value",lis[k])
    k+=1
for j in freq_a:
    print("equal frequency",j,"value",lis[l])
    l+=1

equal width (-11.803, -3.444] value 1
equal width (-11.803, -3.444] value 4
equal width (-11.803, -3.444] value 9
equal width (-3.444, 4.889] value 13
equal width (-3.444, 4.889] value 14
equal width (-3.444, 4.889] value 15
equal width (-3.444, 4.889] value 16
equal width (-3.444, 4.889] value 17
equal width (4.889, 13.222] value 26
equal frequency (-11.779, -1.111] value 1
equal frequency (-11.779, -1.111] value 4
equal frequency (-11.779, -1.111] value 9
equal frequency (-1.111, 2.556] value 13
equal frequency (-1.111, 2.556] value 14
equal frequency (-1.111, 2.556] value 15
equal frequency (2.556, 13.222] value 16
equal frequency (2.556, 13.222] value 17
equal frequency (2.556, 13.222] value 26


**b)**

In [47]:
lis_b = (lis-(np.mean(lis)))/np.std(lis)
wid_b = pd.cut(lis_b,3)
freq_b = pd.qcut(lis_b,3)
k = 0
l = 0
for i in wid_b:
    print("equal width",i,"value",lis[k])
    k+=1
for j in freq_b:
    print("equal frequency",j,"value",lis[l])
    l+=1

equal width (-1.689, -0.493] value 1
equal width (-1.689, -0.493] value 4
equal width (-1.689, -0.493] value 9
equal width (-0.493, 0.7] value 13
equal width (-0.493, 0.7] value 14
equal width (-0.493, 0.7] value 15
equal width (-0.493, 0.7] value 16
equal width (-0.493, 0.7] value 17
equal width (0.7, 1.892] value 26
equal frequency (-1.686, -0.159] value 1
equal frequency (-1.686, -0.159] value 4
equal frequency (-1.686, -0.159] value 9
equal frequency (-0.159, 0.366] value 13
equal frequency (-0.159, 0.366] value 14
equal frequency (-0.159, 0.366] value 15
equal frequency (0.366, 1.892] value 16
equal frequency (0.366, 1.892] value 17
equal frequency (0.366, 1.892] value 26


**c)**

In [45]:
lis_c = np.exp((lis-(np.mean(lis)))/np.std(lis))
wid_c = pd.cut(lis_c,3)
freq_c = pd.qcut(lis_c,3)
k = 0
l = 0
for i in wid_c:
    print("equal width",i,"value",lis[k])
    k+=1
for j in freq_c:
    print("equal frequency",j,"value",lis[l])
    l+=1

equal width (0.179, 2.334] value 1
equal width (0.179, 2.334] value 4
equal width (0.179, 2.334] value 9
equal width (0.179, 2.334] value 13
equal width (0.179, 2.334] value 14
equal width (0.179, 2.334] value 15
equal width (0.179, 2.334] value 16
equal width (0.179, 2.334] value 17
equal width (4.483, 6.633] value 26
equal frequency (0.184, 0.882] value 1
equal frequency (0.184, 0.882] value 4
equal frequency (0.184, 0.882] value 9
equal frequency (0.882, 1.445] value 13
equal frequency (0.882, 1.445] value 14
equal frequency (0.882, 1.445] value 15
equal frequency (1.445, 6.633] value 16
equal frequency (1.445, 6.633] value 17
equal frequency (1.445, 6.633] value 26


✅Answer:

From the above transformations, we can see that, method a and method b does not change how equal width or equal frequency boxing the list variables, but method c change how equal width assign variables to the bins, it was 3,5,1 and now 8,1. I think this is because exponential trnasformation change the "difference multiple" between numbers, what was larger was more influenced by the exponential transformation than the smaller numbers. Thus this is why method c can change the bins when using equal_width method, as it is more sensitive to outliers as we talked on class. For equal frequency, it will always be 3,3,3 as how we set it to be, just like what we say "robust to outliers".

**Question 8**

**a)**

*Approach 2*

In [7]:
w1 = (150+200+280+130+240+110+130+120+180+160)/10

print("6 and 12 with approach 2 has weight",w1)

6 and 12 with approach 2 has weight 170.0


*Approach 3*

In [6]:
male = (150+200+280+130+240)/5
female = (110+130+120+180+160)/5
print("6 with approach 3 has weight", male)
print("12 with approach 3 has weight", female)

6 with approach 3 has weight 200.0
12 with approach 3 has weight 140.0


**b)**

*Approach 1*

In [10]:
wb1 = (150+200+280+130+240+110+130+120+180+160)/10
wb1

170.0

*Approach 2*

In [8]:
wb = (150+200+280+130+240+110+130+120+180+160+w1+w1)/12
wb

170.0

*Approach 3*

In [9]:
wb2 = (150+200+280+130+240+110+130+120+180+160+male+female)/12
wb2

170.0

✅From the above we can see that, all the approaches will result in the same average weight. which is 170.

**c)**

✅Answer:

The "dropna" method will be good when we have a large scale of data size, when one or two missing value does not have a huge inpact on overall average, such as, when we have one minion entries of data, dropping two rows with missing values is not a big deal.